pkg/syncer/syncers/downsyncer.go

```diff
@@ -0,0 +1,181 @@
+
+	downSyncer := DownSyncer{
+		logger:                  logger,
+		upstreamClientFactory:   upstreamClientFactory,
+		downstreamClientFactory: downstreamClientFactory,
+	}
+
+	if err := downSyncer.initializeClients(syncedResources, conversions); err != nil {
+		logger.Error(err, "failed to initialize clients")
+		return &downSyncer, err
+	}
+
+	return &downSyncer, nil
+}
+
+func (ds *DownSyncer) initializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.upstreamClients = map[schema.GroupKind]*Client{}
+	ds.downstreamClients = map[schema.GroupKind]*Client{}
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) ReInitializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) getClients(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) (*Client, *Client, error) {
+	return getClients(resource, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) SyncOne(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.logger.V(3).Info(fmt.Sprintf("downsync '%s'", resourceToString(resource)))
+	upstreamClient, downstreamClient, err := ds.getClients(resource, conversions)
+	if err != nil {
+		ds.logger.Error(err, fmt.Sprintf("failed to get client '%s'", resourceToString(resource)))
+		return err
+	}
+	resourceForUp := convertToUpstream(resource, conversions)
+	ds.logger.V(3).Info(fmt.Sprintf("  get '%s' from upstream", resourceToString(resourceForUp)))
+	upstreamResource, err := upstreamClient.Get(resourceForUp)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			ds.logger.V(3).Info(fmt.Sprintf("  not found '%s' in upstream", resourceToString(resourceForUp)))
+			ds.logger.V(3).Info(fmt.Sprintf("  skip downsync '%s'", resourceToString(resourceForUp)))
+			return nil
+		} else {
+			ds.logger.Error(err, fmt.Sprintf("failed to get resource from upstream '%s'", resourceToString(resourceForUp)))
+			return err
+		}
+	}
+
+	resourceForDown := convertToDownstream(resource, conversions)
+	ds.logger.V(3).Info(fmt.Sprintf("  get '%s' from downstream", resourceToString(resourceForDown)))
+	downstreamResource, err := downstreamClient.Get(resourceForDown)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// create
+			ds.logger.V(3).Info(fmt.Sprintf("  create '%s' in upstream since it's not found", resourceToString(resourceForDown)))
```

Comment:
Isn't the creation being done in the downstream?

Reply:
Yes, correct. Will fix the comment.